IdentityCore/src/validation/MSIDCIAMAuthority.m

```diff
@@ -0,0 +1,187 @@
+}
+
+- (nullable instancetype)initWithURL:(nonnull NSURL *)url
+                      validateFormat:(BOOL)validateFormat
+                           rawTenant:(nullable NSString *)rawTenant
+                             context:(nullable id<MSIDRequestContext>)context
+                               error:(NSError **)error
+{    
+    self = [self initWithURL:url validateFormat:validateFormat context:context error:error];
+    if (self)
+    {
+        if (rawTenant)
+        {
+            if ([self.class isAuthorityFormatValid:url context:context error:nil])
+            {
+                for(int a = 0; a < url.pathComponents.count; a = a + 1 ) {
+                    NSLog(@"value of path component %d: %@\n", a, url.pathComponents[a]);
+                   }
+                _url = [NSURL URLWithString:[NSString stringWithFormat:@"https://%@/%@", [url msidHostWithPortIfNecessary], rawTenant]];
+                _realm = rawTenant;
+            }
+        }
+    }
+
+    return self;
+}
+
+- (nullable instancetype)initWithURL:(NSURL *)url
+                      validateFormat:(BOOL)validateFormat
+                             context:(id<MSIDRequestContext>)context
+                               error:(NSError **)error
+{
+    self = [super initWithURL:url validateFormat:validateFormat context:context error:error];
+    
+    NSArray *hostComponents = [url.msidHostWithPortIfNecessary componentsSeparatedByString:@"."];
+
+    //If we have the URL tenant.ciamlogin.com
+    if (url.pathComponents.count == 0)
+    {
+        url = [url URLByAppendingPathComponent:hostComponents[0]];
+        url = [NSURL URLWithString:[url.absoluteString stringByAppendingString:@".onmicrosoftonline.com"]];
+    }
+    
+    if (self)
+    {
+        _url = url;
+        _url = [MSIDAADAuthority normalizedAuthorityUrl:url context:context error:error];
+        if (!_url) return nil;
+        self.url = url;
+    }
+    
+    return self;
+}
+    
+- (instancetype)initWithURL:(NSURL *)url
+                    context:(id<MSIDRequestContext>)context
+                      error:(NSError **)error
+{
+    return [self initWithURL:url validateFormat:YES context:context error:error];
+}
+
++ (BOOL)isAuthorityFormatValid:(NSURL *)url
+                       context:(id<MSIDRequestContext>)context
+                         error:(NSError **)error
+{
+    if (![super isAuthorityFormatValid:url context:context error:error]) return NO;
+    
+    NSArray *hostComponents = [url.msidHostWithPortIfNecessary componentsSeparatedByString:@"."];
+
+    if (hostComponents.count < 3)
+    {
+        if (error)
+        {
+            *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorInternal, @"CIAM authority should have at least 3 segments in the path (i.e. https://<tenant>.<host>.<com>...)", nil, nil, nil, context.correlationId, nil, YES);
+        }
+        
+        return NO;
+    }
+    
+    if (![hostComponents[1]  isEqual: @"ciamlogin"])
```

Comment:
would a malformed url (for example "a..b") crash this check because of null?

Reply:
Also case sensitive. I think we have some string utility classes to safely compare string values.